File: backend/app/models/team.py
from sqlalchemy import Column, Integer, String, Text, ForeignKey, DateTime, Boolean, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base

class Team(Base):
    __tablename__ = "teams"
    
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    hackathon_id = Column(Integer, ForeignKey("hackathons.id"), nullable=False)
    
    # Team Details
    max_members = Column(Integer, default=4)
    current_members = Column(Integer, default=1)
    required_skills = Column(JSON, default=list)  # ["Python", "React"]
    looking_for_roles = Column(JSON, default=list)  # ["Frontend Dev", "Designer"]
    
    # Project Info
    project_idea = Column(Text, nullable=True)
    tech_stack = Column(JSON, default=list)  # ["React", "Node.js", "MongoDB"]
    github_repo = Column(String(255), nullable=True)
    
    # Status
    is_public = Column(Boolean, default=True)  # Can others see and join
    is_complete = Column(Boolean, default=False)  # Team is full
    status = Column(String(50), default="forming")  # forming, working, submitted
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Relationships
    hackathon = relationship("Hackathon", back_populates="teams")
    members = relationship("TeamMember", back_populates="team")

class TeamMember(Base):
    __tablename__ = "team_members"
    
    id = Column(Integer, primary_key=True, index=True)
    team_id = Column(Integer, ForeignKey("teams.id"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    
    role = Column(String(100), nullable=True)  # "Team Lead", "Frontend Dev"
    is_leader = Column(Boolean, default=False)
    joined_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    team = relationship("Team", back_populates="members")
    user = relationship("User", back_populates="team_memberships")

File: backend/src/models/Team.js
const mongoose = require('mongoose');

const teamSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true,
    maxlength: 100
  },
  description: {
    type: String,
    required: true,
    maxlength: 500
  },
  hackathon: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Hackathon',
    required: true
  },
  leader: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  members: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    joinedAt: {
      type: Date,
      default: Date.now
    },
    role: {
      type: String,
      enum: ['leader', 'member'],
      default: 'member'
    },
    skills: [String] // Skills they bring to the team
  }],
  requiredSkills: [{
    skill: String,
    priority: {
      type: String,
      enum: ['high', 'medium', 'low'],
      default: 'medium'
    }
  }],
  maxMembers: {
    type: Number,
    default: 4,
    min: 2,
    max: 10
  },
  status: {
    type: String,
    enum: ['open', 'full', 'closed'],
    default: 'open'
  },
  tags: [String], // Tech stack, domain, etc.
  lookingFor: {
    type: String,
    maxlength: 300
  },
  contactInfo: {
    discord: String,
    slack: String,
    email: String,
    telegram: String
  },
  projectIdea: {
    title: String,
    description: String,
    techStack: [String],
    githubRepo: String
  },
  isPublic: {
    type: Boolean,
    default: true
  },
  applicationRequired: {
    type: Boolean,
    default: false
  },
  applications: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    message: String,
    status: {
      type: String,
      enum: ['pending', 'accepted', 'rejected'],
      default: 'pending'
    },
    appliedAt: {
      type: Date,
      default: Date.now
    }
  }],
  achievements: [{
    title: String,
    description: String,
    date: Date
  }]
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for current member count
teamSchema.virtual('currentMembers').get(function() {
  return this.members.length;
});

// Virtual for available spots
teamSchema.virtual('availableSpots').get(function() {
  return this.maxMembers - this.members.length;
});

// Virtual to check if team is full
teamSchema.virtual('isFull').get(function() {
  return this.members.length >= this.maxMembers;
});

// Indexes for efficient queries
teamSchema.index({ hackathon: 1, status: 1 });
teamSchema.index({ leader: 1 });
teamSchema.index({ 'members.user': 1 });
teamSchema.index({ tags: 1 });
teamSchema.index({ requiredSkills: 1 });

// Pre-save middleware to update status
teamSchema.pre('save', function(next) {
  if (this.members.length >= this.maxMembers) {
    this.status = 'full';
  } else if (this.status === 'full' && this.members.length < this.maxMembers) {
    this.status = 'open';
  }
  next();
});

// Methods
teamSchema.methods.addMember = function(userId, skills = []) {
  if (this.isFull) {
    throw new Error('Team is full');
  }
  
  const isAlreadyMember = this.members.some(member => 
    member.user.toString() === userId.toString()
  );
  
  if (isAlreadyMember) {
    throw new Error('User is already a team member');
  }
  
  this.members.push({
    user: userId,
    skills: skills,
    role: 'member'
  });
  
  return this.save();
};

teamSchema.methods.removeMember = function(userId) {
  if (this.leader.toString() === userId.toString()) {
    throw new Error('Team leader cannot be removed');
  }
  
  this.members = this.members.filter(member => 
    member.user.toString() !== userId.toString()
  );
  
  return this.save();
};

teamSchema.methods.isMember = function(userId) {
  return this.members.some(member => 
    member.user.toString() === userId.toString()
  ) || this.leader.toString() === userId.toString();
};

teamSchema.methods.canUserJoin = function(userId) {
  return !this.isFull && !this.isMember(userId) && this.status === 'open';
};

// Static methods
teamSchema.statics.findByHackathon = function(hackathonId, filters = {}) {
  const query = { hackathon: hackathonId };
  
  if (filters.status) query.status = filters.status;
  if (filters.tags) query.tags = { $in: filters.tags };
  if (filters.availableOnly) query.status = 'open';
  
  return this.find(query)
    .populate('leader', 'name email profilePicture skills')
    .populate('members.user', 'name email profilePicture skills')
    .populate('hackathon', 'name startDate endDate')
    .sort({ createdAt: -1 });
};

teamSchema.statics.findUserTeams = function(userId) {
  return this.find({
    $or: [
      { leader: userId },
      { 'members.user': userId }
    ]
  })
  .populate('hackathon', 'name startDate endDate')
  .populate('leader', 'name email profilePicture')
  .populate('members.user', 'name email profilePicture skills');
};


const Team = mongoose.model('Team', teamSchema);

module.exports = Team;